vidyut-lipi/src/lib.rs

```diff
@@ -1,118 +1,551 @@
+	let tok: String = node.tok_val.clone();
+	// This could possibly use a Rc, combined with a RefCell
+	// for suffix. But did not work on initial try.
+	// So may be update it later.
+	// ToDo -- SanB 1/9/23
+	let nnode = node.clone();
+	self.enc_to_tok.insert(insstr, node);
+	self.tok_to_enc.insert(tok, nnode);
+    }
+
+    fn update_enc_suffix(&mut self, sref: String, suff: bool) {
+	let rval = self.enc_to_tok.get_mut(&sref);
+	if rval.is_some() {
+	    (*self.enc_to_tok.get_mut(&sref).unwrap()).suffix = suff;
+	}
+    }
+
+    
+    // To handle multi-character encoding that does not have a valid prefix
+    // we will need to insert these prefixes, but set them as invalid.
+    // For example to handle "Rrt" - in an itrans stream - Rrt is not
+    // valid encoding - but Rri is. So Rr is a valid prefix, so R, and Rr
+    // Need to be inserted as prefix but with no valid token translaton. 
+    fn add_prefixes(&mut self, sref: &mut String) {
+	while sref.chars().count() > 1 {
+	    // First check whether prefix already present.
+	    sref.pop();
+	    
+	    let cnode = self.enc_to_tok.get(sref);
+	    if cnode.is_none() {
+		let node = Node::init(sref.clone(), "".to_string(), "".to_string(),
+				      TOK_TRANSPARENT_TYPE, true, IMPLICIT_UNKNOWN);
+		self.insert_enc_to_tok(node);
+	    } else {
+		
+	    }
+	}
+    }
+    
+
+    fn get_enc_map(&self, sref: String) -> Option<&Node> {
+	let rval = self.enc_to_tok.get(&sref);
+	return rval;
+    }
+
+    #[allow(dead_code)]
+    fn get_tok_map(&self, sref: String) -> Option<&Node> {
+	return self.tok_to_enc.get(&sref)
+    }
+
+    pub fn update_string_with_token_with_implicit(&self, intok: String, instr: &mut String,
+						  imp: u8,prev:u8) {
+
+	let tnode = self.get_tok_map(intok);
+	if tnode.is_some() {
+	    let w = tnode.unwrap();
+	    match w.tok_type {
+		TOK_IS_CONSONANT => {
+		    instr.push_str(&w.mnem);
+		    if w.implicit != imp {
+			if imp == IMPLICIT_VIRAMA {
+			    let vnode = self.get_tok_map("TOK_VOWEL_SIGN_VIRAMA".to_string());
+			    if vnode.is_some() {
+				instr.push_str(&vnode.unwrap().mnem);
+			    }			    
+			}
+		    }
+		},
+		TOK_IS_VOWEL => {
+		    if prev == TOK_IS_CONSONANT {
+			    instr.pop(); // pop the associated virama
+			    instr.push_str(&w.assoc);
+		    } else {
+			instr.push_str(&w.mnem);
+		    }
+		}
+		TOK_LEXICAL_BREAK => {
+		    
+		},
+		_ => {
+		    instr.push_str(&w.mnem);
+		}
+	    }
+
+	    
+	    if w.implicit != imp && w.tok_type == TOK_IS_CONSONANT {
+		if imp == IMPLICIT_VIRAMA {
+
+		}
+	    }
+
+	    
+	}
+    }
+    
+    // debug print of the translationmap. 
+    pub fn print_transmap(&self) {
+	for (key, value) in &(self.enc_to_tok) {
+	    println!("Key: {}, {},{},{},{},{}", key, value.mnem,
+		     value.tok_val,
+		     value.suffix,
+		     value.tok_type,
+		     value.implicit);
+	}
+    }
+    
+}
+
+
+pub fn generate_map<P: AsRef<Path>>(path: P, transmap: &mut TranslationMap ) -> &mut TranslationMap {
```

Comment:
rather than pass in a new map, let's create and return the map here:

```
pub fn generate_map<P: AsRef<Path>>(path: P) -> TranslationMap {
...
}
```